### PR TITLE
refactor: specify conditions that affect liquidationCollateral

### PR DESCRIPTION
**Pull request by gregdhill**
Signed-off-by: Gregory Hill <gregorydhill@outlook.com>

### CODE REVIEW

File: interbtc-spec/docs/source/spec/vault-registry.rst [hunk 1]
```diff
@@ -749,12 +759,16 @@ Specification
...
-   * The ``oldVault``'s ``toBeRedeemedTokens`` and ``issuedTokens`` MUST be decreased by the amount ``tokens``.
-   * The ``oldVault``'s collateral free balance MUST be increased by ``tokens / toBeRedeemed``.
-* If ``oldVault`` *is* liquidated, the liquidation vault's ``toBeRedeemedTokens`` and ``issuedTokens`` are decrease by the amount ``tokens``.
-* If ``newVault`` is *not* liquidated, its ``toBeIssuedTokens`` is decreased by ``tokens``, while its ``issuedTokens`` is increased by the same amount.
-* If ``newVault`` *is* liquidated, the liquidation vault's  ``toBeIssuedTokens`` is decreased by ``tokens``, while its ``issuedTokens`` is increased by the same amount.
+* If the ``oldVault`` *is* liquidated:
+
+   * The amount ``toBeReleased`` is calculated as ``(oldVault.liquidatedCollateral * tokens) / toBeRedeemedTokens``.
+   * The ``oldVault``'s ``liquidatedCollateral`` MUST decrease by ``toBeReleased``.
+   * Function :ref:`reward_depositStake` MUST complete successfully - parameterized by ``oldVault`` and ``toBeReleased``.
```
Comment (sander2) [inline interbtc-spec/docs/source/spec/vault-registry.rst:766]:
```suggestion
   * Function :ref:`staking_depositStake` MUST complete successfully - parameterized by ``oldVault``, ``oldVault`` and ``toBeReleased``.
```

File: interbtc-spec/docs/source/spec/vault-registry.rst [hunk 2]
```diff
@@ -823,10 +837,29 @@ Specification
 
 *Postconditions*
 
-* Function :ref:`reward_withdrawStake` MUST complete successfully - parameterized by ``vault`` and ``vault.issuedTokens``.
-* The liquidation vault's ``issuedTokens``, ``toBeIssuedTokens`` and ``toBeRedeemedTokens`` MUST be increased by the respective amounts in the vault.
-* The vault's ``issuedTokens`` and ``toBeIssuedTokens`` MUST be set to 0.
-* Collateral MUST be moved from the vault to the liquidation vault: an amount of ``confiscatedCollateral - confiscatedCollateral * (toBeRedeemedTokens / (toBeIssuedTokens + issuedTokens))`` is moved, where ``confiscatedCollateral`` is the minimum of the vault's ``backingCollateral`` (as calculated via :ref:`computeStakeAtIndex`) and ``SecureCollateralThreshold`` times the equivalent worth of the amount of tokens it is backing.
+* ``usedCollateral`` MUST be calculated as ``exchangeRate * (issuedTokens + toBeIssuedTokens)) * secureCollateralThreshold``.
+* ``usedCollateral`` MUST be set to ``backingCollateral`` if ``backingCollateral < usedCollateral``.
+* ``usedTokens`` MUST be calculated as ``issuedTokens - toBeIssuedTokens``.
```
Comment (sander2) [inline interbtc-spec/docs/source/spec/vault-registry.rst:842]:
```suggestion
* ``usedTokens`` MUST be calculated as ``issuedTokens + toBeIssuedTokens``.
```

File: interbtc-spec/docs/source/spec/vault-registry.rst [hunk 3]
```diff
@@ -823,10 +837,29 @@ Specification
...
+* ``toBeLiquidated`` MUST be calculated as ``(usedCollateral * (usedTokens - toBeRedeemedTokens)) / usedTokens``.
+* ``remainingCollateral`` MUST be calculated as ``usedCollateral - toBeLiquidated``.
```
Comment (sander2) [inline interbtc-spec/docs/source/spec/vault-registry.rst:844]:
```suggestion
* ``remainingCollateral`` MUST be calculated as ``max(0, usedCollateral - toBeLiquidated)``.
```

File: interbtc-spec/docs/source/spec/vault-registry.rst [hunk 4]
```diff
@@ -823,10 +837,29 @@ Specification
...
+* Function :ref:`reward_withdrawStake` MUST complete successfully - parameterized by ``vault`` and ``issuedTokens``.
+* Function :ref:`staking_withdrawStake` MUST complete successfully - parameterized by ``vault`` and ``remainingCollateral``.
```
Comment (sander2) [inline interbtc-spec/docs/source/spec/vault-registry.rst:846]: I think this is wrong. The call to `slash_to_liquidation_vault` is with `to_be_liquidated` as amount, which then does some calculation on.

As a side note, having the naming mismatch makes it a lot harder to check correctness..

**Review by sander2 — CHANGES_REQUESTED** [review]

File: interbtc-spec/docs/source/spec/vault-registry.rst [hunk 5]
```diff
@@ -580,18 +580,28 @@ One of:
...
+   * If ``premium > 0``, then ``premium`` is transferred from the vault's collateral to the redeemer.
    * Function :ref:`reward_withdrawStake` MUST complete successfully - parameterized by ``vaultId`` and ``tokens``.
-* If the vault *is* liquidated, then the liquidation vault's ``toBeRedeemedTokens`` MUST be decreased by ``tokens``, and its ``issuedTokens`` MUST increase by the same amount. The ``RedeemTokensLiquidatedVault`` event is emitted.
+   * The redeemer's free collateral balance MUST increase by ``premium``.
+
+* If the vault *is* liquidated:
+
+   * The amount ``toBeReleased`` is calculated as ``(liquidatedCollateral * tokens) / toBeRedeemedTokens``.
+   * The vault's ``liquidatedCollateral`` MUST decrease by ``toBeReleased``.
+   * Function :ref:`reward_depositStake` MUST complete successfully - parameterized by ``vaultId`` and ``toBeReleased``.
```
[the remark this answers is not in the document]
Comment (gregdhill) [reply interbtc-spec/docs/source/spec/vault-registry.rst:601]: This is already correct, issued interBTC is the reward stake.

File: interbtc-spec/docs/source/spec/vault-registry.rst (hunk 1 above)
Comment (gregdhill) [reply interbtc-spec/docs/source/spec/vault-registry.rst:766]: Same as above, this affects interBTC reward stake.

File: interbtc-spec/docs/source/spec/vault-registry.rst (hunk 4 above)
Comment (gregdhill) [reply interbtc-spec/docs/source/spec/vault-registry.rst:846]: I'm not sure exactly what you think is wrong, could you clarify?

Comment (sander2) [reply interbtc-spec/docs/source/spec/vault-registry.rst:845]: The argument to `staking_withdrawStake`. I don't think it's called with `remainingCollateral`, but due to the confusing naming I am not 100% sure.

Comment (gregdhill) [reply interbtc-spec/docs/source/spec/vault-registry.rst:845]: I'll merge and resolve after the audit fixes - will harmonize the naming to make this easier to check.

**Pull request merged**